internal/allocator/allocator.go

```
@@ -194,113 +187,118 @@ func (a *Allocator) Unassign(svc string) bool {
+			delete(a.sharingKeyForIP, ip.String())
+		}
+		a.poolIPsInUse[al.pool][ip.String()]--
+		if a.poolIPsInUse[al.pool][ip.String()] == 0 {
+			// Explicitly delete unused IPs from the pool, so that len()
+			// is an accurate count of IPs in use.
+			delete(a.poolIPsInUse[al.pool], ip.String())
 		}
-		delete(a.portsInUse[al.ip.String()], port)
-	}
-	delete(a.servicesOnIP[al.ip.String()], svc)
-	if len(a.portsInUse[al.ip.String()]) == 0 {
-		delete(a.portsInUse, al.ip.String())
-		delete(a.sharingKeyForIP, al.ip.String())
-	}
-	a.poolIPsInUse[al.pool][al.ip.String()]--
-	if a.poolIPsInUse[al.pool][al.ip.String()] == 0 {
-		// Explicitly delete unused IPs from the pool, so that len()
-		// is an accurate count of IPs in use.
-		delete(a.poolIPsInUse[al.pool], al.ip.String())
 	}
 	stats.poolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPsInUse[al.pool])))
 	return true
 }
 
-func cidrIsIPv6(cidr *net.IPNet) bool {
-	return cidr.IP.To4() == nil
-}
-func ipIsIPv6(ip net.IP) bool {
-	return ip.To4() == nil
-}
-
 // AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svc string, isIPv6 bool, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, error) {
+func (a *Allocator) AllocateFromPool(svc string, serviceIPFamily ipfamily.IPFamily, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
 	if alloc := a.allocated[svc]; alloc != nil {
 		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "ipFamily" is an immutable field in services.
-		if isIPv6 != ipIsIPv6(alloc.ip) {
-			return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ip.String())
+		// This "should-not-happen" since the "serviceIPFamily" is an immutable field in services.
+		if allocipsFamily, err := ipfamily.IPFamilyForAddressesIPs(alloc.ips); err != nil || allocipsFamily != serviceIPFamily {
```

nit: allocIPsFamily
I'd also call `IPFamilyForAddressesIPs` out of the if and check the error first, the family after in two separate ifs. Here we are swallowing the err and the if is a bit convoluted.